climada/engine/test/test_impact.py

```diff
@@ -69,6 +69,134 @@ def test_from_eih_pass(self):
+        with self.assertRaises(ValueError) as cm:
+            imp3.crs = 'OTHER'
+            Impact.concat([imp1, imp2, imp3])
+        self.assertIn("The impacts have different exposure crs", str(cm.exception))
+        #reset crs
+        imp3.crs = imp2.crs
+
+        #test total exposure value
+        with self.assertRaises(ValueError) as cm:
+            imp3.tot_value = 1
+            Impact.concat([imp1, imp2, imp3])
+        self.assertIn("The impacts have different total exposure values", str(cm.exception))
+        # reset exposure value
+        imp3.tot_value = imp2.tot_value
+
+        #test exposure coordinates
+        with self.assertRaises(ValueError) as cm:
+            imp3.coord_exp[0][0] = 0
+            Impact.concat([imp1, imp2, imp3])
+        self.assertIn("The impacts have different exposure coordinates", str(cm.exception))
+        # reset exposure coordinates
+        imp3.coord_exp = imp2.coord_exp
+
+    def test_check_event_ids_and_imp_mat(self):
+        """test event_id checks and impact matrices check"""
+
+        # get test data
+        ent = Entity.from_excel(ENT_DEMO_TODAY)
+        ent.check()
+        hazard = Hazard.from_hdf5(HAZ_TEST_TC)
+
+        # create hazard subsets
+        date0 = hazard.date[0]
+        date1 = hazard.date[int(np.floor(len(hazard.date) / 3))]
+        date2 = hazard.date[int(np.floor(len(hazard.date) / 3)) * 2]
+        date3 = hazard.date[-1]
+        haz_subset1 = hazard.select(date=(date0, date1 - 1))
+        haz_subset2 = hazard.select(date=(date1, date2 - 1))
+        haz_subset3 = hazard.select(date=(date2, date3))
+
+        # compute impacts separately
+        imp1 = ImpactCalc(ent.exposures, ent.impact_funcs, haz_subset1).impact()
+        imp2 = ImpactCalc(ent.exposures, ent.impact_funcs, haz_subset2).impact()
+        imp3 = ImpactCalc(ent.exposures, ent.impact_funcs, haz_subset3).impact()
+
+        # test event ids
+        with self.assertRaises(ValueError) as cm:
+            imp3.event_id[0] = imp2.event_id[-1]
+            Impact.concat([imp1, imp2, imp3])
+        self.assertIn("Duplicate event IDs found", str(cm.exception))
+        #reset event ids
+        imp3.event_id[0] = imp3.event_id[0] + 1
+
+        # test impact matrices
+        with self.assertRaises(ValueError) as cm:
+            imp3.imp_mat = None
+            Impact.concat([imp1, imp2, imp3])
+        self.assertIn("Passed an impact object with an undefined impact matrix", str(cm.exception))
+
+        with self.assertRaises(ValueError) as cm:
+            imp3.imp_mat = sparse.csr_matrix(np.empty((0, 0)))
+            Impact.concat([imp1, imp2, imp3])
+        self.assertIn("Impact matrices do not have the same number of exposure points", str(cm.exception))
+
+
+    def test_results(self):
+        """test results of impact.concat"""
+
+        # get test data
+        ent = Entity.from_excel(ENT_DEMO_TODAY)
+        ent.check()
+        hazard = Hazard.from_hdf5(HAZ_TEST_TC)
```

Comment:
These are unit tests, so I think you should create your own impact objects with controlled data instead of some random input. You do not need hazards or exposures for that. Simply fill your impact objects with data you know and test the data in the resulting impact object explicitly. Note that `assert_array_equal` also compares the array shapes/lengths. Please write tests allong these lines (with the other Impact attributes added as well):
```python
impact_1 = Impact(frequency=[0.1])
impact_2 = Impact(frequency=[0.2])

impact = Impact.concat(impact_1, impact_2)
np.testing.assert_array_equal(impact.frequency, [0.1, 0.2])
```
This goes for all tests. If you want to check that concatenation works with "real world" examples, this should go into an integration test